File: Kotlin/src/main/kotlin/parrot/Parrot.kt
package parrot

import kotlin.math.min

open class Parrot(
    private val type: ParrotTypeEnum,
    val numberOfCoconuts: Int,
    val voltage: Double,
    val isNailed: Boolean
) {

    open val speed: Double
        get() = baseSpeed

    val loadFactor: Double
        get() = 9.0

    val baseSpeed: Double
        get() = 12.0

    fun getBaseSpeed(voltage: Double): Double = min(24.0, voltage * baseSpeed)
}

File: TypeScript/jest/src/parrot.ts
export class Parrot {
    constructor(
                public numberOfCoconuts: number,
                public voltage: number,
                public isNailed: boolean) {
    }

    public getBaseSpeed(): number {
        return 12;
    }

}

export class EuropeanParrot extends Parrot {
    constructor() {
        super(0, 0, false);
    }

    getSpeed() {
        return 12;
    }
}

export class AfricanParrot extends Parrot {
    constructor( numberOfCoconuts: number,
                 voltage: number) {
        super(numberOfCoconuts, voltage, false);
    }

    getSpeed(): number {
        return Math.max(0, this.getBaseSpeed() - this.getLoadFactor() * this.numberOfCoconuts);
    }

    getLoadFactor(): number {
        return 9;
    }
}

export class NorwegianBlueParrot extends Parrot {
    constructor( numberOfCoconuts: number,
                 voltage: number,
                 isNailed: boolean) {
        super( numberOfCoconuts, voltage, isNailed);
    }

    getSpeed(): number {
        return (this.isNailed) ? 0 : this.getBaseSpeedWithVoltage(this.voltage);
    }

    getBaseSpeedWithVoltage(voltage: number): number {
        return Math.min(24, voltage * this.getBaseSpeed());
    }
}
